File: python/ibrawls_rl/protocol.py
"""Binary wire protocol client — the Python mirror of ``src/sim/server/protocol.ts``.

Frames are a uint32 **big-endian** length prefix followed by the payload. The payload's
first byte is an opcode; step tensors are little-endian (``<f4`` / ``<i4``). JSON is used
only for the one-time HELLO handshake.
"""
from __future__ import annotations

import json
import struct
from dataclasses import dataclass
from typing import IO

import numpy as np

# Opcodes — keep in sync with protocol.ts OPCODE.
HELLO = 0
RESET = 1
STEP = 2
CLOSE = 3
STATE = 4  # request: opcode + uint32 world index; response: JSON state snapshot


def write_frame(stream: IO[bytes], payload: bytes) -> None:
    """Write a length-prefixed frame and flush."""
    stream.write(struct.pack(">I", len(payload)))
    stream.write(payload)
    stream.flush()


def read_frame(stream: IO[bytes]) -> bytes:
    """Block until a full frame is read; return its payload."""
    header = _read_exact(stream, 4)
    (length,) = struct.unpack(">I", header)
    return _read_exact(stream, length)


def _read_exact(stream: IO[bytes], n: int) -> bytes:
    chunks = []
    remaining = n
    while remaining > 0:
        chunk = stream.read(remaining)
        if not chunk:
            raise EOFError("sim server closed the pipe")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def hello_request(config: dict) -> bytes:
    return bytes([HELLO]) + json.dumps(config).encode("utf-8")


def parse_hello_response(payload: bytes) -> dict:
    assert payload[0] == HELLO, f"expected HELLO, got opcode {payload[0]}"
    return json.loads(payload[1:].decode("utf-8"))


def reset_request() -> bytes:
    return bytes([RESET])


def close_request() -> bytes:
    return bytes([CLOSE])


def state_request(world_index: int = 0) -> bytes:
    """Ask for one world's render-ready JSON snapshot (the Watch tab's feed)."""
    return bytes([STATE]) + struct.pack("<I", int(world_index))


def parse_state_response(payload: bytes) -> dict:
    return json.loads(payload.decode("utf-8"))


def step_request(actions: np.ndarray) -> bytes:
    """Pack an int32 action block (any shape) into a STEP payload."""
    a = np.ascontiguousarray(actions, dtype="<i4")
    return bytes([STEP]) + a.tobytes()


@dataclass
class StepResponse:
    obs: np.ndarray
    reward: np.ndarray
    done: np.ndarray
    truncated: np.ndarray
    # agent-flat-index -> terminal observation (obs_dim,), for done agents.
    terminal_obs: dict
    # Aggregate reward components for the worker step, in header["rewardComponentKeys"] order.
    reward_components: np.ndarray


def parse_step_response(
    payload: bytes,
    n_agents: int,
    obs_dim: int,
    reward_component_count: int = 0,
) -> StepResponse:
    """Slice a step-response payload into obs / reward / done / truncated + terminal obs.

    Layout mirrors ``buildStepResponse`` in protocol.ts:
      obs(f32) · reward(f32) · done(u8) · truncated(u8) · nTerminal(u32) ·
      [ idx(u32) · termObs(f32, obs_dim) ] × nTerminal
    """
    off = 0
    obs = np.frombuffer(payload, dtype="<f4", count=n_agents * obs_dim, offset=off)
    off += n_agents * obs_dim * 4
    reward = np.frombuffer(payload, dtype="<f4", count=n_agents, offset=off)
    off += n_agents * 4
    done = np.frombuffer(payload, dtype=np.uint8, count=n_agents, offset=off)
    off += n_agents
    truncated = np.frombuffer(payload, dtype=np.uint8, count=n_agents, offset=off)
    off += n_agents

    (n_terminal,) = struct.unpack_from("<I", payload, off)
    off += 4
    terminal_obs: dict = {}
    for _ in range(n_terminal):
        (idx,) = struct.unpack_from("<I", payload, off)
        off += 4
        term = np.frombuffer(payload, dtype="<f4", count=obs_dim, offset=off).copy()
        off += obs_dim * 4
        terminal_obs[idx] = term

    reward_components = np.zeros((0,), dtype=np.float32)
    if off + 4 <= len(payload):
        (encoded_count,) = struct.unpack_from("<I", payload, off)
        off += 4
        count = reward_component_count or encoded_count
        reward_components = np.zeros((count,), dtype=np.float32)
        readable = min(count, encoded_count)
        if readable:
            reward_components[:readable] = np.frombuffer(
                payload, dtype="<f4", count=readable, offset=off
            )

    return StepResponse(
        obs=obs.reshape(n_agents, obs_dim).copy(),
        reward=reward.copy(),
        done=done.copy(),
        truncated=truncated.copy(),
        terminal_obs=terminal_obs,
        reward_components=reward_components.copy(),
    )


def parse_obs_only(payload: bytes, n_agents: int, obs_dim: int) -> np.ndarray:
    """Parse a RESET response (obs block only)."""
    obs = np.frombuffer(payload, dtype="<f4", count=n_agents * obs_dim, offset=0)
    return obs.reshape(n_agents, obs_dim).copy()

File: src/sim/server/protocol.ts
/**
 * Length-prefixed binary wire protocol between the Python trainer and this Node vec-env.
 * JSON is used **only** for the one-time handshake (config in, header out); every step
 * exchanges raw little-endian typed-array bytes — no per-step JSON on the hot path.
 *
 * Frame = uint32 big-endian length prefix, then `length` bytes of payload.
 *   - Request payload: 1 opcode byte, then opcode-specific data.
 *       RESET: just the opcode.
 *       STEP : opcode + an Int32 action block (numEnvs·numAgents·actDim, little-endian).
 *       CLOSE: just the opcode.
 *   - Step response: raw concat of obs(Float32) + reward(Float32) + done(Uint8) blocks.
 *   - Reset response: the obs(Float32) block.
 * Sizes are fixed by the handshake header, so the receiver slices by known lengths.
 */

export const OPCODE = {
  HELLO: 0, // JSON config frame (request) / JSON header frame (response)
  RESET: 1,
  STEP: 2,
  CLOSE: 3,
} as const;

export type Opcode = (typeof OPCODE)[keyof typeof OPCODE];

const LE = true; // little-endian on the wire (matches numpy '<f4' / '<i4')

/** Wrap a payload in a 4-byte big-endian length prefix. */
export function writeFrame(payload: Uint8Array): Uint8Array {
  const out = new Uint8Array(4 + payload.length);
  new DataView(out.buffer).setUint32(0, payload.length, false);
  out.set(payload, 4);
  return out;
}

/**
 * Streaming frame decoder. Push raw socket/stdin chunks; pull complete payloads. Handles
 * frames split across or coalesced within chunks.
 */
export class FrameDecoder {
  private buf: Uint8Array = new Uint8Array(0);

  push(chunk: Uint8Array): void {
    if (this.buf.length === 0) {
      this.buf = chunk.slice();
    } else {
      const merged = new Uint8Array(this.buf.length + chunk.length);
      merged.set(this.buf, 0);
      merged.set(chunk, this.buf.length);
      this.buf = merged;
    }
  }

  /** Next complete payload, or null if a full frame isn't buffered yet. */
  next(): Uint8Array | null {
    if (this.buf.length < 4) return null;
    const len = new DataView(this.buf.buffer, this.buf.byteOffset, 4).getUint32(0, false);
    if (this.buf.length < 4 + len) return null;
    const payload = this.buf.slice(4, 4 + len);
    this.buf = this.buf.slice(4 + len);
    return payload;
  }
}

export function opcodeOf(payload: Uint8Array): number {
  return payload[0];
}

/** Build a request payload for an opcode with no data (RESET / CLOSE). */
export function buildSimpleRequest(opcode: Opcode): Uint8Array {
  return new Uint8Array([opcode]);
}

/** Build a STEP request payload from an Int32 action block. */
export function buildStepRequest(actions: Int32Array): Uint8Array {
  const out = new Uint8Array(1 + actions.length * 4);
  out[0] = OPCODE.STEP;
  const dv = new DataView(out.buffer);
  for (let i = 0; i < actions.length; i++) dv.setInt32(1 + i * 4, actions[i], LE);
  return out;
}

/** Parse a STEP request payload's action block into an Int32Array of length `count`. */
export function parseStepRequest(payload: Uint8Array, count: number): Int32Array {
  const dv = new DataView(payload.buffer, payload.byteOffset + 1, payload.length - 1);
  const out = new Int32Array(count);
  for (let i = 0; i < count; i++) out[i] = dv.getInt32(i * 4, LE);
  return out;
}

/** Raw little-endian bytes of a Float32Array (copied so the source buffer can be reused). */
export function f32Bytes(a: Float32Array): Uint8Array {
  const out = new Uint8Array(a.length * 4);
  const dv = new DataView(out.buffer);
  for (let i = 0; i < a.length; i++) dv.setFloat32(i * 4, a[i], LE);
  return out;
}

/** Decode little-endian Float32 bytes back into a Float32Array (test/round-trip helper). */
export function bytesToF32(bytes: Uint8Array, count: number): Float32Array {
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const out = new Float32Array(count);
  for (let i = 0; i < count; i++) out[i] = dv.getFloat32(i * 4, LE);
  return out;
}

/**
 * Step-response payload:
 *   obs(f32, n·obsDim) · reward(f32, n) · done(u8, n) · truncated(u8, n) ·
 *   nTerminal(u32) · [ idx(u32) · termObs(f32, obsDim) ] × nTerminal
 * where n = numEnvs·numAgents. Terminal observations are sent for every done agent
 * (Gymnasium convention) so the trainer can bootstrap truncations.
 */
export function buildStepResponse(
  obs: Float32Array,
  reward: Float32Array,
  done: Uint8Array,
  truncated: Uint8Array,
  terminalObs: (Float32Array | null)[],
  obsDim: number
): Uint8Array {
  const obsB = f32Bytes(obs);
  const rewB = f32Bytes(reward);
  const n = reward.length;

  const terminalIdx: number[] = [];
  for (let i = 0; i < terminalObs.length; i++) if (terminalObs[i]) terminalIdx.push(i);
  const termBlockLen = 4 + terminalIdx.length * (4 + obsDim * 4);

  const out = new Uint8Array(obsB.length + rewB.length + n + n + termBlockLen);
  let off = 0;
  out.set(obsB, off); off += obsB.length;
  out.set(rewB, off); off += rewB.length;
  out.set(done, off); off += n;
  out.set(truncated, off); off += n;

  const dv = new DataView(out.buffer);
  dv.setUint32(off, terminalIdx.length, LE); off += 4;
  for (const idx of terminalIdx) {
    dv.setUint32(off, idx, LE); off += 4;
    const term = terminalObs[idx] as Float32Array;
    for (let j = 0; j < obsDim; j++) { dv.setFloat32(off, term[j], LE); off += 4; }
  }
  return out;
}

export interface StepResponse {
  obs: Float32Array;
  reward: Float32Array;
  done: Uint8Array;
  truncated: Uint8Array;
  /** agent-flat-index -> terminal observation, for done agents. */
  terminalObs: Map<number, Float32Array>;
}

/** Split a step-response payload back into its blocks given `n` agents and `obsDim`. */
export function parseStepResponse(payload: Uint8Array, n: number, obsDim: number): StepResponse {
  let off = 0;
  const obs = bytesToF32(payload.subarray(off, off + n * obsDim * 4), n * obsDim); off += n * obsDim * 4;
  const reward = bytesToF32(payload.subarray(off, off + n * 4), n); off += n * 4;
  const done = payload.subarray(off, off + n).slice(); off += n;
  const truncated = payload.subarray(off, off + n).slice(); off += n;

  const dv = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const nTerminal = dv.getUint32(off, LE); off += 4;
  const terminalObs = new Map<number, Float32Array>();
  for (let k = 0; k < nTerminal; k++) {
    const idx = dv.getUint32(off, LE); off += 4;
    const term = new Float32Array(obsDim);
    for (let j = 0; j < obsDim; j++) { term[j] = dv.getFloat32(off, LE); off += 4; }
    terminalObs.set(idx, term);
  }
  return { obs, reward, done, truncated, terminalObs };
}
